sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -699,6 +699,24 @@ object CollapseProject extends Rule[LogicalPlan] {
         agg.copy(aggregateExpressions = buildCleanedProjectList(
           p.projectList, agg.aggregateExpressions))
       }
+    case p1 @ Project(_, g @ GlobalLimit(_, l @ LocalLimit(_, p2: Project))) =>
```

Comment:
Sorry to be late for the review. I have 2 concerns about this optimization:
1. if `p2` outputs one column, and `p1` outputs 1000 columns, then pushing down `p1` through limit operator would increase the data size to be shuffled.
2. if `p1` has an expensive expression like UDF, pushing `p1` through limit operator means the expensive expression will be executed a lot more times.

Do we have a general rule to justify the benefit of pushing down the project operator?

Reply:
Thank you for review, @gatorsmile and @cloud-fan . I got it. I'll narrow down with `if isRenaming(l1, l2) =>` for these cases, too.

Reply:
I believe case 1 can be handled with `isRenaming`, but I'm not sure how case 2 can be handled.

Case 2 actually means that we can't push down project through operators that will reduce the numRows, e.g. limit, sample, etc.

Reply:
@cloud-fan . `isRenaming` use `semanticEquals`. Case 2 will be prevented. I'll make a PR soon.